src/shared/image_utils.js

```diff
@@ -13,30 +13,42 @@
  * limitations under the License.
  */
 
+import { FeatureTest } from "./util.js";
+
 function applyMaskImageData({
   src,
   srcPos = 0,
   dest,
-  destPos = 3,
+  destPos = 0,
   width,
   height,
   inverseDecode = false,
 }) {
-  const srcLength = src.byteLength;
-  const zeroMapping = inverseDecode ? 0 : 255;
-  const oneMapping = inverseDecode ? 255 : 0;
+  const opaque = FeatureTest.isLittleEndian ? 0xff000000 : 0x000000ff;
+  const [zeroMapping, oneMapping] = !inverseDecode ? [opaque, 0] : [0, opaque];
+  const widthInSource = width >> 3;
+  const widthRemainder = width & 7;
+  dest = new Uint32Array(dest.buffer);
 
-  for (let j = 0; j < height; j++) {
-    let elem,
-      mask = 0;
-    for (let k = 0; k < width; k++) {
-      if (mask === 0) {
-        elem = srcPos < srcLength ? src[srcPos++] : 255;
-        mask = 128;
-      }
-      dest[destPos] = elem & mask ? oneMapping : zeroMapping;
-      destPos += 4;
-      mask >>= 1;
+  for (let i = 0; i < height; i++) {
+    for (const max = srcPos + widthInSource; srcPos < max; srcPos++) {
+      const elem = srcPos < src.length ? src[srcPos] : 255;
```

Comment:
Previously we computed this *once* at the start of the function, which seems more efficient overall. Hence, please re-introduce the `srcLength` variable and use that here and below.